x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationService.java

```diff
@@ -0,0 +1,173 @@
+        DiscoveryNode masterNode = currentState.nodes().getMasterNode();
+        if (masterNode == null) {
+            logger.warn(
+                "[{}] no master known for allocation state update [{}]",
+                request.getModelId(),
+                request.getRoutingState().getState()
+            );
+            waitForNewMasterAndRetry(observer, UpdateTrainedModelAllocationStateAction.INSTANCE, request, listener, changePredicate);
+            return;
+        }
+        client.execute(UpdateTrainedModelAllocationStateAction.INSTANCE, request, ActionListener.wrap(listener::onResponse, failure -> {
+            if (isMasterChannelException(failure)) {
+                logger.info(
+                    "[{}] master channel exception will retry on new master node for allocation state update [{}]",
+                    request.getModelId(),
+                    request.getRoutingState().getState()
+                );
+                waitForNewMasterAndRetry(observer, UpdateTrainedModelAllocationStateAction.INSTANCE, request, listener, changePredicate);
+                return;
+            }
+            listener.onFailure(failure);
+        }));
+    }
+
+    public void createNewModelAllocation(
+        StartTrainedModelDeploymentAction.TaskParams taskParams,
+        ActionListener<CreateTrainedModelAllocationAction.Response> listener
+    ) {
+        client.execute(CreateTrainedModelAllocationAction.INSTANCE, new CreateTrainedModelAllocationAction.Request(taskParams), listener);
+    }
+
+    public void deleteModelAllocation(String modelId, ActionListener<AcknowledgedResponse> listener) {
+        client.execute(DeleteTrainedModelAllocationAction.INSTANCE, new DeleteTrainedModelAllocationAction.Request(modelId), listener);
+    }
+
+    public void waitForAllocationCondition(
+        final String modelId,
+        final Predicate<TrainedModelAllocation> predicate,
+        final @Nullable TimeValue timeout,
+        final WaitForAllocationListener listener
+    ) {
+        final Predicate<ClusterState> clusterStatePredicate = clusterState -> predicate.test(
+            TrainedModelAllocationMetadata.allocationForModelId(clusterState, modelId).orElse(null)
+        );
+
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, timeout, logger, threadPool.getThreadContext());
+        final ClusterState clusterState = observer.setAndGetObservedState();
+        if (clusterStatePredicate.test(clusterState)) {
+            listener.onResponse(TrainedModelAllocationMetadata.allocationForModelId(clusterState, modelId).orElse(null));
+        } else {
+            observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    listener.onResponse(TrainedModelAllocationMetadata.allocationForModelId(clusterState, modelId).orElse(null));
```

Comment:
```suggestion
                    listener.onResponse(TrainedModelAllocationMetadata.allocationForModelId(state, modelId).orElse(null));
```

Shouldn't the clusterstate in the method parameter be used here?